### ISSUE TITLE
CMake enhancement

### ISSUE THREAD

**Issue by promgamer** [private-octopus/picoquic#310]
We are trying to use PICOQUIC as a 3rdparty dependency on a CMake project. We found some problems that we solve by applying patches after checkout.

- Relative paths (with "../../../").
- Include picotls easily
- Duplicated declarations on CMakeLists.txt

Will you accept a CMAKE enhancement that solves some of those problems?

**Comment by bkchr** [comment]
Just make the pull request and than let's see :)

**Comment by huitema** [comment]
@promgamer as Bastian said, yes, we are absolutely accepting patches.

**Comment by huitema** [comment]
PR #420 removes the usage of relative paths (../.., etc.) in header declarations and also in test file configuration. Tests and demo programs can be run from any location, provided that the user specifies the path to the Picoquic solution in a command line argument, e.g. "picoquic_ct -S ~/picoquic". @promgamer, does that resolve your "../../.." comment?

**Comment by huitema** [comment]
There are some remaining direct references in `CMakeList.txt`, such as:
```
FIND_LIBRARY(PTLS_CORE picotls-core PATH ../picotls)
MESSAGE(STATUS "Found picotls-core at : ${PTLS_CORE} " )

FIND_LIBRARY(PTLS_MINICRYPTO picotls-minicrypto PATH ../picotls)
MESSAGE(STATUS "Found picotls-crypto at : ${PTLS_MINICRYPTO} " )

FIND_LIBRARY(PTLS_OPENSSL picotls-openssl PATH ../picotls)
MESSAGE(STATUS "Found picotls-openssl at : ${PTLS_OPENSSL} " )
```
Any suggestion on how to replace them?

**Comment by promgamer** [comment]
Well, I see at least 3 options here:

1. Use `find_library` on picoquic and provide a `FindPicoTLS.cmake` with picoTLS.
2. Include picoTLS as a git submodule of picoquic
3. Use Cmake variables like `-DPICOTLS_DIR` that we can define to point it to picoTLS

1 and 3 "decouple" more picoquic and picotls by being more flexible with what you can reference (either via CMake's find package or via CMake variables. But it may be harder to execute and maintain.
2 seems easier execute and maintain, but it couples picoquic's repository to picotls repository (which doesn't seem a problem to me, to be honest :smile: )

**Comment by huitema** [comment]
I don't like making Picotls into a submodule, because Picotls is a library by itself, that could also be used with other protocols, say DTLS. Plus, eventually, we should be able to install Picotls as a packaged library, much like we install OpenSSL. The easiest might be:

1) Use a Cmake variable
2) If that variable is not defined, assume default = ../picotls
3) Use the variable in the FIND_LIBRARY commands, and also to define the include paths.

How would that sound?

**Comment by promgamer** [comment]
Sounds good. I would make it incremental. Start with step n. 1 and 2. And then, possibly iterate to step n. 3

**Comment by huitema** [comment]
@alagoutte any idea on how to remove the dependency on the location of picotls library?

**Comment by alagoutte** [comment]
Sorry not a expert on packaging but i think, the solution 1 is better ( @Lekensteyn may be a idea ?)

**Comment by Lekensteyn** [comment]
The current `find_library` invocation is probably broken, it uses the `PATH` option which is not documented. So `FIND_LIBRARY(PTLS_CORE picotls-core PATH ../picotls)` will look for `picotls-core`, `PATH` and `../picotls`.

I think it should be:
```cmake
find_library(PTLS_CORE picotls-core PATHS ${CMAKE_BINARY_DIR}/../picotls)
```

(That assumes that the build directory of picotls is next to the build directory of picoquic.) To avoid hardcoding guesses like that, the whole `PATHS ...` part could be dropped. In that case you will have to add the picotls build dir to the `CMAKE_LIBRARY_PATH` list variable though.

Preferably a FindPTLS.cmake module should be created which allows the include dir and library path to be overridden by the user (when using `find_package`).

Is picotls versioned? If so, then the minimum version requirement could also be exposed via `find_package`. FindPTLS.cmake would then have to extract the version from somewhere (a header file?) for comparison.

**Comment by huitema** [comment]
How do I go about creating "_a FindPTLS.cmake module_" ?

**Comment by huitema** [comment]
Thanks. So the idea is to have:

1) a cmake script ./cmake/FindPTLS.cmake that installs variables for the include path and the libraries

2) a call to find_package(PTLS) in the CMakeLists.txt

The FindPTLS script would use commands like find_path or find_library, maybe with HINTS such as HINTS ${CMAKE_BINARY_DIR}/../picotls

If that's the idea, I can start experimenting...

**Comment by Lekensteyn** [comment]
Yes, that should work.

**Comment by huitema** [comment]
I now have PR #479 that adds FindPTLS.cmake. I tested that it odes work on my Ubuntu VM when PTLS is located at ../picotls, so at least we do not have a migration. But this is very bare bone. @Lekensteyn, does it feel right? @promgamer, do you believe that will work for you?

**Comment by Lekensteyn** [comment]
I've left some inline comments to PR #479. The hard-coded hints are not ideal, but they should only be used as fallback when the cache variables (e.g. `PTLS_CORE_LIBRARY`) is not set and if the library was not found in other paths (e.g. list variable `CMAKE_LIBRARY_PATH`).

**Comment by huitema** [comment]
Fixed in PR #479
